### PR TITLE
NFC Documentation

### PR DESCRIPTION
**Pull request by donatieng**
This is the associated docs PR for https://github.com/ARMmbed/mbed-os/pull/7822

### CODE REVIEW

File: docs/reference/api/connectivity/nfc/NFC.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    virtual void on_nfc_initiator_discovered(const SharedPtr< mbed::nfc::NFCRemoteInitiator> &nfc_initiator) {
+        printf("Remote inititator detected\r\n");
+        nfc_initiator->set_delegate(nfc_initiator_delegate_ptr);
+        nfc_initiator->connect(); // Connect to the initiator
+    }
+};
+```
+
+#### NFC controller
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFC.md]: If these are separate APIs, they should each get their own page and description.

File: docs/reference/api/connectivity/nfc/NFC.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+#### NDEF API
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/FIXME)
+
+### Design
+
+A detailed design document is available [within the Mbed OS source tree](https://github.com/ARMmbed/mbed-os/docs/design-documents/nfc/nfc_design.md). It details the rationale behind the API design.
+
+[explorenfc_nucleo]: explorenfc_nucleo.jpg
+[n_mark]: n_mark.png
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFC.md]: Query: Do we have an example created for NFC?

File: docs/reference/api/connectivity/nfc/NFC.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+NFC offers three modes:
+
+1. NFC card emulation.
+1. NFC reader or writer.
+1. NFC peer to peer.
+
+Mbed OS supports the card emulation mode, either through the use of a controller or NFC EEPROM.
+
+NFC support is available for the NXP PN512 transceiver. In terms of NFC EEPROMs, a driver for the ST M24SR EEPROM series is available - we also provide a [porting guide](https://os.mbed.com/contributing/connectivity/NFCEEPROMDriver.html) to add support for other NFC EEPROMs.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFC.md]: Let's remove the first part of this, so what we support doesn't quickly become out of date.

File: docs/reference/api/connectivity/nfc/NFC.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+Most NFC controllers implement many modes of NFC and offer flexibility; however, they are typically more expensive than NFC EEPROMs.
+
+NFC controllers typically support a large subset (if not all) of NFC Forum functionality, can act as either initiators or targets and are driven by an external stack. In Mbed OS, this means that you can generate NDEF messages "on demand" when the initiator reads them.
+
+NFC EEPROMs behave like NFC tags whose memory can either be addressed through a wired (such as I2C) or NFC air interface. One benefit is that these work autonomously from a microcontroller. Some of them can also use the NFC field as a power source if they are powered off at the time.
+
+### Getting started with NFC and example
+
+NFC examples are available on [GitHub](https://github.com/ARMmbed/mbed-os-example-nfc) and demonstrate how to create NFC tags that you can be read from and write to using a phone.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFC.md]: Note to self: Delete extra word.

File: docs/reference/api/connectivity/nfc/NFC.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/FIXME)
+
+#### NDEF API
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/FIXME)
+
+### Design
+
+A detailed design document is available [within the Mbed OS source tree](https://github.com/ARMmbed/mbed-os/docs/design-documents/nfc/nfc_design.md). It details the rationale behind the API design.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFC.md]: This link is broken, and I can't seem to find where to go to replace it. Could you please point me to the right one?

**Comment by MarceloSalazar** [comment]
The docs is looking good.
I feel some of the content in the intro belongs to the technology section?

Also, would be nice to update the photo "explorenfc_nucleo.jpg" as it looks a bit messy with many cables and can't see much details on the NFC shield.

File: docs/reference/api/connectivity/nfc/NFC.md (hunk 5 above)
Comment (donatieng) [reply docs/reference/api/connectivity/nfc/NFC.md]: That's this one - the link above will become valid once the PR is merged in :)
https://github.com/donatieng/mbed-os/blob/nfc-spec/docs/design-documents/nfc/nfc_design.md

Comment (MarceloSalazar) [reply docs/reference/api/connectivity/nfc/NFC.md]: There is **really** good content here. I'd love to see this is the technology section.

File: docs/reference/api/connectivity/nfc/NDEF.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    } delegate;
+
+    parser.set_delegate(&delegate);
+    parser.parse(buffer);
+}
+```
+
+An application can extend capabilities of `SimpleMessageParser` by adding new record parsers (`mbed::nfc::ndef::RecordParser`) at runtime.
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/FIXME)
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NDEF.md]: Note to self: Add correct link after code merges.

File: docs/reference/api/connectivity/nfc/NDEF.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    Text text(Text::UTF8, span_from_cstr("en-US"), span_from_cstr("Mbed website"));
+
+    uri.append_as_record(builder);
+    text.append_as_record(builder, /* last record */ true);
+
+    return builder.get_message().size();
+}
+```
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/FIXME)
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NDEF.md]: Note to self: Add correct link after code merges.

File: docs/reference/api/connectivity/nfc/NFCController.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
+## NFC controller
+
+To use an NFC controller, you must initiate the instance with a driver instance, an event queue and a scratch buffer for NDEF messages.
+
+### NFC controller class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/FIXME)
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFCController.md]: Note to self: Add correct link after code merges.

File: docs/reference/api/connectivity/nfc/NFCEEPROM.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
+## NFC EEPROM
+
+To use an NFC controller, you must initiate the instance with a driver instance, an event queue and a scratch buffer for NDEF messages.
+
+### NFC EEPROM class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/FIXME)
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFCEEPROM.md]: Note to self: Add correct link after code merges.

File: docs/reference/api/connectivity/nfc/NDEF.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
+### NDEF API
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NDEF.md]: I love this file. Nice work.

File: docs/reference/api/connectivity/nfc/NFCController.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
+## NFC controller
+
+To use an NFC controller, you must initiate the instance with a driver instance, an event queue and a scratch buffer for NDEF messages.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFCController.md]: This tells us how to use it but doesn't say what it is. Could we add a sentence saying what this is and when/why you'd want to use it?

File: docs/reference/api/connectivity/nfc/NFCEEPROM.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
+## NFC EEPROM
+
+To use an NFC controller, you must initiate the instance with a driver instance, an event queue and a scratch buffer for NDEF messages.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/nfc/NFCEEPROM.md]: I'm not sure how this sentence about an NFC controller ties in to this page about NFC EEPROM. Could we say what NFC EEPROM is, what it stands for and when/why you'd want to use it?

**Comment by AnotherButler** [comment]
Note to self: Fix Doxy links after code merges, and upload images after doc merges.

**Comment by AnotherButler** [comment]
@donatieng Could you also resolve the merge conflicts when you're finished with your changes?

File: docs/reference/api/connectivity/nfc/NDEF.md (hunk 6 above)
Comment (MarceloSalazar) [reply docs/reference/api/connectivity/nfc/NDEF.md]: Do you think it makes sense to add this code at it is, or should we let users import it?
Maybe this code is already available in the mbed-os-example-nfc application?

File: docs/reference/api/connectivity/nfc/NDEF.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+using mbed::nfc::ndef::RecordID;
+using mbed::nfc::ndef::MessageParser;
+using mbed::nfc::ndef::common::Text;
+using mbed::nfc::ndef::common::URI;
+using mbed::nfc::ndef::common::Mime;
+
+using mbed::nfc::ndef::common::SimpleMessageParser;
+
+void parse_ndef_message(const Span<const uint8_t> &buffer) {
+    SimpleMessageParser parser;
```
Comment (MarceloSalazar) [inline docs/reference/api/connectivity/nfc/NDEF.md]: What's the purpose of these functions?
Are they included in Mbed OS, or something that a developer will need to copy & paste in their application?

File: docs/reference/api/connectivity/nfc/NDEF.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+```
+#include "nfc/ndef/MessageBuilder.h"
+
+using mbed::nfc::ndef::MessageBuilder;
+using mbed::nfc::ndef::common::Text;
+using mbed::nfc::ndef::common::URI;
+
+size_t build_ndef_message(const Span<uint8_t> &buffer) {
+    MessageBuilder builder(buffer);
```
Comment (MarceloSalazar) [inline docs/reference/api/connectivity/nfc/NDEF.md]: Same question for this function.
Is this included in Mbed OS or should the developer include this code in their app?

File: docs/reference/api/connectivity/nfc/NDEF.md (hunk 10 above)
Comment (donatieng) [reply docs/reference/api/connectivity/nfc/NDEF.md]: That's Copyright @pan- :)

File: docs/reference/api/connectivity/nfc/NDEF.md (hunk 13 above)
Comment (pan-) [reply docs/reference/api/connectivity/nfc/NDEF.md]: The line above the code it is indicated that `mbed::nfc::ndef::common::SimpleMessageParser` is part of the API provided. 

The code just demonstrate how to use this class to parse an NDEF message.

File: docs/reference/api/connectivity/nfc/NDEF.md (hunk 14 above)
Comment (pan-) [reply docs/reference/api/connectivity/nfc/NDEF.md]: Same as above, it is explicitly that `mbed::nfc::ndef::MessageBuilder` is part of the API mbed OS provide. 

It just demonstrate how an application can create an NDEF message.

File: docs/reference/api/connectivity/nfc/NDEF.md (hunk 6 above)
Comment (pan-) [reply docs/reference/api/connectivity/nfc/NDEF.md]: We do not have any parsing in our examples yet plus 60 lines of well focused code that demonstrate the API are different than an example which just solves a specific use case with the whole NFC API.

Technology examples and API demonstrations are different beasts.

**Comment by AnotherButler** [comment]
@donatieng Thanks for making those changes 👍

**Comment by AnotherButler** [comment]
Note to self: Add pages to .json after merging.

**Comment by AnotherButler** [comment]
@donatieng I can't find the APIs in the Doxygen. Could you please help me?

**Pull request merged**